### PR TITLE
Feat/condition groups visualization

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
## Types of PR

- [x] New Features
- [ ] Upgrade of existing features
- [ ] Bugfix

## Description

- Added visualization feature of status of the `ConditionGroup` defined in the scenario file.

[condition_group_visualization.webm](https://github.com/tier4/scenario_simulator_v2/assets/32741405/4a1e77db-c3f0-4862-80b1-01681f5aa87a)

- Minor update rviz config file
  - disable simulator internal information's visualization
  - modify order of the visualization feature
- Add document of Visualization feature's explanation

### CODE REVIEW

**Comment by kyoichi-sugahara** [comment]
@kosuke55 
Could you take a look at rviz config modification and add some information if you think it's necessary?

**Comment by kosuke55** [comment]
@kyoichi-sugahara thanks seems good for rviz config to me.

**Comment by hakuturu583** [comment]
I think this pull request is very helpful for Autoware developers.
So please update [ReleaseNotes.md](https://github.com/tier4/scenario_simulator_v2/blob/master/docs/ReleaseNotes.md)

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+  YAML::Node data;
+  try {
+    data = YAML::Load(msg_ptr->data);
+  } catch (const std::exception & e) {
+    throw std::runtime_error(std::string("Failed to load YAML: ") + e.what());
+  }
+
+  int unnamed_event_counter = 1;
+  condition_groups_collection_ptr_ = std::make_shared<ConditionGroupsCollection>();
```
Comment (hakuturu583) [inline openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:207]:
```suggestion
  condition_groups_collection_ptr_->clear();
```

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  try {
+    event_name = event_node["name"].as<std::string>();
+  } catch (const YAML::BadConversion & e) {
+    event_name = "";
+  }
+  if (event_name.empty()) {
+    event_name = "ConditionGroup" + std::to_string(counter++);
+  }
+
+  ConditionGroups condition_groups_msg;
```
Comment (hakuturu583) [inline openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:248]: `ConditionGroups` is not a ros2/protobuf message.
I suggest to rename variable.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

**Comment by hakuturu583** [comment]
As the Conditions are displayed in parallel, it is difficult to recognise where each Condition comes from.

It would be appreciated if you could indicate where the Condition is tied to, such as `ConditionGroup "hoge" (StartTrigger of Event "fuga")`.

**Comment by HansRobo** [comment]
@kyoichi-sugahara Could you merge latest master into this branch for X2 regression test?

File: openscenario/openscenario_visualization/package.xml [hunk 3]
```diff
@@ -5,9 +5,11 @@
   <version>0.7.0</version>
   <description>Visualization tools for simulation results</description>
   <maintainer email="masaya.kataoka@tier4.jp">Masaya Kataoka</maintainer>
+  <maintainer email="kyoichi.sugahara@tier4.jp">Kyoichi Sugahara</maintainer>
   <license>Apache License 2.0</license>
 
   <buildtool_depend>ament_cmake</buildtool_depend>
+  <buildtool_depend>ament_cmake_auto</buildtool_depend>
   <buildtool_depend>pkg-config</buildtool_depend>
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/package.xml:14]: added with https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/aa7e7fb09a7b4b22dfbed9d740fcf3422b8a48f8

File: openscenario/openscenario_visualization/include/openscenario_visualization_condition_groups_plugin/jsk_overlay_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+#include <QImage>
+#include <QVariant>
+#include <rclcpp/rclcpp.hpp>
+#include <rviz_rendering/render_system.hpp>
+
+namespace jsk_rviz_plugins
+{
+class OverlayObject;
+
+class ScopedPixelBuffer
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/include/openscenario_visualization_condition_groups_plugin/jsk_overlay_utils.hpp:96]: added in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/378a73fca9c98511eed45c9a2bc0b2db13c21894

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+  property_top_->setMin(0);
+
+  property_length_ = new rviz_common::properties::IntProperty(
+    "Length", length, "Length of the plotter window", this, SLOT(updateVisualization()), this);
+  property_length_->setMin(10);
+  property_width_ = new rviz_common::properties::IntProperty(
+    "Width", width, "Width of the plotter window", this, SLOT(updateVisualization()), this);
+  property_width_->setMin(10);
+  property_value_scale_ = new rviz_common::properties::FloatProperty(
+    "Value Scale", text_size, "Value scale", this, SLOT(updateVisualization()), this);
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp]: added in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/72dc5dc7218e6a8d7b547fdba65a59682ffe20a7

File: docs/user_guide/VisualizingInternalInformation.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+- **Current Evaluation**: Current achievement status
+- **Current Value**: The fulfillment status of the condition expressed in true/false
+- **Type**: Types of conditions, such as `ReachPositionCondition`, `CollisionCondition`, etc.
+
+To visualize the achievement status, check the `ConditionGroup` checkbox as demonstrated in the image below. 
+![condition group visualization](../image/condition_group.png "condition group visualization")") 
+
+If an `Event name` is defined in the `ConditionGroup`, that name will be reflected. If left blank, names like `ConditionGroup1`, `ConditionGroup2`, `ConditionGroup3`, etc. are automatically assigned. However, this information alone doesn't clarify if each `ConditionGroup` is linked to success or failure conditions. Therefore, when visualizing with rviz and you want this information to be apparent, we recommend explicitly defining the `Event name` as `SuccessCondition` or similar when creating the scenario file.
+
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply docs/user_guide/VisualizingInternalInformation.md:55]: added in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/378a73fca9c98511eed45c9a2bc0b2db13c21894

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+namespace openscenario_visualization
+{
+VisualizationConditionGroupsDisplay::VisualizationConditionGroupsDisplay()
+: condition_groups_collection_ptr_(std::make_shared<std::vector<ConditionGroups>>())
+{
+  // Get screen info of default display
+  const Screen * screen_info = DefaultScreenOfDisplay(XOpenDisplay(NULL));
+
+  // Fixed height for a 4k resolution screen
+  constexpr float hight_4k = 2160.0;
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:37]: confirmed the plugin works with non-4K display
image

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+  const float text_size = scale * 35.0;
+
+  // Define initial value of left edge position of condition results panel
+  const int left = 0;
+
+  // Define initial value of top edge position of condition results panel
+  const int top = static_cast<int>(std::round(450 * scale));
+
+  // Define initial value of horizontal length of condition results panel
+  const int length = static_cast<int>(std::round(2000 * scale));
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:51]: The multiplication of 2000 is used to define the initial horizontal length and width of the panel. This value is not necessarily the one set from the Width property; it's a predefined value for creating a panel of a certain size. The use of 2000 as a base value suggests that the panel is designed to be significantly wide to accommodate detailed information.
The actual width of the panel in the application is scaled based on the user's screen resolution compared to a 4K resolution. This scaling ensures that the panel maintains a consistent size and appearance across different screen resolutions.

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+  // Calculate scale based on current screen height and 4k height
+  const float scale = static_cast<float>(screen_info->height) / hight_4k;
+
+  const float text_size = scale * 35.0;
+
+  // Define initial value of left edge position of condition results panel
+  const int left = 0;
+
+  // Define initial value of top edge position of condition results panel
+  const int top = static_cast<int>(std::round(450 * scale));
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:48]: The multiplication of 450 sets the initial top edge position of the condition results panel. Like the width and length, this is a predefined base value, not necessarily linked to any value set from the Top property.
The purpose of this calculation is to position the top edge of the panel at an appropriate place on the screen, again scaling according to screen resolution to maintain a consistent look across different devices.

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+  // Get screen info of default display
+  const Screen * screen_info = DefaultScreenOfDisplay(XOpenDisplay(NULL));
+
+  // Fixed height for a 4k resolution screen
+  constexpr float hight_4k = 2160.0;
+
+  // Calculate scale based on current screen height and 4k height
+  const float scale = static_cast<float>(screen_info->height) / hight_4k;
+
+  const float text_size = scale * 35.0;
```
[the remark this answers is not in the document]
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:42]: The multiplication of 35.0 is used to determine the initial text size in the panel. This value is a base size that is then scaled according to the screen resolution.
The 'Value Scale' property allows users to adjust the scaling factor of this text size later on, but the initial value of 35.0 is set to ensure a default size that is likely suitable for most screens. The scaling factor adjusts this size to ensure readability across various resolutions.

**Comment by kyoichi-sugahara** [comment]
> As the Conditions are displayed in parallel, it is difficult to recognise where each Condition comes from.
> 
> It would be appreciated if you could indicate where the Condition is tied to, such as `ConditionGroup "hoge" (StartTrigger of Event "fuga")`.

@hakuturu583 
I agree with your idea but difficult for me to find the best way to visualize information and get the information based on the conclusion.
So can we discuss how to deal with this internally?

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp (hunk 2 above)
Comment (kyoichi-sugahara) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp]: changed in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/acdbc5a3017e305b20f8cf82013859e88642059f

File: docs/ReleaseNotes.md [hunk 11]
```diff
@@ -18,7 +18,7 @@ Major Changes :race_car: :red_car: :blue_car:
 Bug Fixes:bug:
 
 | Feature                                  | Brief summary                                                                                                                                 | Category                                           | Pull request                                                      | Contributor                                   |
-| ---------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------- | ----------------------------------------------------------------- | --------------------------------------------- |
+|------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------|-------------------------------------------------------------------|-----------------------------------------------|
```
Comment (HansRobo) [inline docs/ReleaseNotes.md:21]: Please don't create diffs that depend on the formatter settings only in your specific environments.
```suggestion
| ---------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------- | ----------------------------------------------------------------- | --------------------------------------------- |
```

Comment (kyoichi-sugahara) [reply docs/ReleaseNotes.md:21]: ops thank you for the comment!
I will format with default setting at the end, sorry
(I used formatter setting for autoware : just execuse :bow: :bow: :bow:

Comment (hakuturu583) [reply docs/ReleaseNotes.md:21]: @kyoichi-sugahara You can modify it by using commit suggestion.

**Comment by kyoichi-sugahara** [comment]
converts to draft for now :bow:

**Comment by kyoichi-sugahara** [comment]
@hakuturu583 
> It would be appreciated if you could indicate where the Condition is tied to, such as ConditionGroup "hoge" (StartTrigger of Event "fuga").

your suggestion is adding StartTrigger of Event "fuga" in each condition's description?
Like 
startTrigger: "fuga"
image

If so is this necessary?
Because if conditions are in the same conditionGroup, startTrigger of each condition is same, and it's same with Condition Groups Name.

Now i feel Condition Group Name should be renamed to event name though...

**Comment by hakuturu583** [comment]
I think the string is yaml, so string like below is better.

```
ConditionGroup:
  Name: name is not defined
  StartTrigger: hogehoge
  Conditions:
    - CurrentEvaluation: hoge
      CurrentValue: false
      Type: CollisionCondition
```

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:42]: Could you add this information as doxygen comment?

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:48]: Could you add this information as doxygen comment?

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:51]: Could you add this information as doxygen comment?

**Review by hakuturu583 — CHANGES_REQUESTED** [review]
Please also update release note.

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:59]: Fix in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/b1f8c0ae3c30c3228ec47ec6f41e4e3ebadff540

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:62]: Fixed in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/afe59e556be802378c620f4dabd82280939c3145

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:50]: Fixed in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/b1f8c0ae3c30c3228ec47ec6f41e4e3ebadff540

File: openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+  const int left = 0;
+
+  // Define initial value of top edge position of condition results panel
+  const int top = static_cast<int>(std::round(450 * scale));
+
+  // Define initial value of horizontal length of condition results panel
+  const int length = static_cast<int>(std::round(2000 * scale));
+
+  // Define initial value of width of condition results panel
+  const int width = static_cast<int>(std::round(2000 * scale));
```
[the remark this answers is not in the document]
Comment (hakuturu583) [reply openscenario/openscenario_visualization/src/openscenario_visualization_condition_groups_plugin/openscenario_visualization_condition_groups_plugin.cpp:76]: Fixed in https://github.com/tier4/scenario_simulator_v2/pull/1033/commits/aa1b875fc257ed87c95a973a04907644b4a6feab

**Review by hakuturu583 — APPROVED** [review]

**Comment by hakuturu583** [comment]
review from @hakuturu583 was passed.
Try starting evaluator.

**Comment by hakuturu583** [comment]
Regression does not found in Autoware Evaluator.

**Pull request merged**